app/src/hid_keyboard.c

```diff
@@ -0,0 +1,347 @@
+ * The former byte is item type prefix, we only use short items here.
+ * For how to calculate an item, read 6.2.2 Report Descriptor in
+ * <https://www.usb.org/sites/default/files/hid1_11.pdf>.
+ * For Consumer Page tags, see 15 Consumer Page (0x0C) in
+ * <https://usb.org/sites/default/files/hut1_2.pdf>.
+ */
+/**
+ * You can dump your device's report descriptor with
+ * `sudo usbhid-dump -m vid:pid -e descriptor`.
+ * Change `vid:pid` to your device's vendor ID and product ID.
+ */
+unsigned char kb_report_desc_buffer[]  = {
+    // Usage Page (Generic Desktop)
+    0x05, 0x01,
+    // Usage (Keyboard)
+    0x09, 0x06,
+
+    // Collection (Application)
+    0xA1, 0x01,
+    // Report ID (1)
+    0x85, 0x01,
+
+    // Usage Page (Keyboard)
+    0x05, 0x07,
+    // Usage Minimum (224)
+    0x19, 0xE0,
+     // Usage Maximum (231)
+    0x29, 0xE7,
+    // Logical Minimum (0)
+    0x15, 0x00,
+    // Logical Maximum (1)
+    0x25, 0x01,
+    // Report Size (1)
+    0x75, 0x01,
+    // Report Count (8)
+    0x95, 0x08,
+    // Input (Data, Variable, Absolute): Modifier byte
+    0x81, 0x02,
+
+    // Report Size (8)
+    0x75, 0x08,
+    // Report Count (1)
+    0x95, 0x01,
+    // Input (Constant): Reserved byte
+    0x81, 0x01,
+
+    // Usage Page (LEDs)
+    0x05, 0x08,
+    // Usage Minimum (1)
+    0x19, 0x01,
+    // Usage Maximum (5)
+    0x29, 0x05,
+    // Report Size (1)
+    0x75, 0x01,
+    // Report Count (5)
+    0x95, 0x05,
+    // Output (Data, Variable, Absolute): LED report
+    0x91, 0x02,
+
+    // Report Size (3)
+    0x75, 0x03,
+    // Report Count (1)
+    0x95, 0x01,
+    // Output (Constant): LED report padding
+    0x91, 0x01,
+
+    // Usage Page (Key Codes)
+    0x05, 0x07,
+    // Usage Minimum (0)
+    0x19, 0x00,
+    // Usage Maximum (101)
+    0x29, HID_KEYBOARD_KEYS - 1,
+    // Logical Minimum (0)
+    0x15, 0x00,
+    // Logical Maximum(101)
+    0x25, HID_KEYBOARD_KEYS - 1,
+    // Report Size (8)
+    0x75, 0x08,
+    // Report Count (6)
+    0x95, HID_KEYBOARD_MAX_KEYS,
+    // Input (Data, Array): Keys
+    0x81, 0x00,
+
+    // End Collection
+    0xC0,
+
+    // Usage Page (Consumer)
+    0x05, 0x0C,
+    // Usage (Consumer Control)
+    0x09, 0x01,
+
+    // Collection (Application)
+    0xA1, 0x01,
+    // Report ID (2)
+    0x85, 0x02,
+
+    // Usage Page (Consumer)
+    0x05, 0x0C,
+    // Usage (Scan Next Track)
+    0x09, 0xB5,
+    // Usage (Scan Previous Track)
+    0x09, 0xB6,
+    // Usage (Stop)
+    0x09, 0xB7,
+    // Usage (Eject)
+    0x09, 0xB8,
+    // Usage (Play/Pause)
+    0x09, 0xCD,
+    // Usage (Mute)
+    0x09, 0xE2,
+    // Usage (Volume Increment)
+    0x09, 0xE9,
+    // Usage (Volume Decrement)
+    0x09, 0xEA,
+    // Logical Minimum (0)
+    0x15, 0x00,
+    // Logical Maximum (1)
+    0x25, 0x01,
+    // Report Size (1)
+    0x75, 0x01,
+    // Report Count (8)
+    0x95, 0x08,
+    // Input (Data, Array)
+    0x81, 0x02,
```

Comment:
The media keys should not be forwarded (see my other comment), so this could be removed too.

(IIUC, this also removes the need for handling `Report ID`s.)

Reply:
Oh, unless they are generated from the scrcpy shortcuts (e.g. Alt + up could trigger the HID media volume up). :+1:

Reply:
Could be removed (except we want to implement shortcuts with them).

I kept them just in case.

(Removing the Report ID byte might change a lot code, so I suggest not to remove Report ID, it should be fine I think.)